### ISSUE TITLE
stdlib unittest change breaks a trial test

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2605]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2605|
|Type|defect|
|Created|2007-04-25 11:48:48Z|
|Branch|https://github.com/twisted/twisted/tree/testcase-comparison-2605|

The change in question is here:
http://svn.python.org/view/python/trunk/Lib/unittest.py?rev=54199&r1=42115&r2=54199

The test broken is twisted.trial.test.test_class.FactoryCounting.test_runMultipleCopies. It doesn't seem to have over consequences.

It seems it happens after the add of !__eq!__ to unittest.TestCase. As we use a set to hold instances of a TestCase, a second instance of TestCase with the same attributes can't be added (because now it's equal to the first instance).

I'm not really sure if the change in Python is not a regression, but we can probably provide a workaround for trial.

<details><summary>Searchable metadata</summary>

```
trac-id__2605 2605
type__defect defect
reporter__therve therve
priority__normal normal
milestone__Python_2_6 Python-2.6
branch__branches_testcase_comparison_2605 branches/testcase-comparison-2605
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1177501728000000 1177501728000000
changetime__1225490851000000 1225490851000000
version__None None
owner__ 
cc__exarkun cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

http://mail.python.org/pipermail/python-dev/2007-October/074973.html

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25198]) Branching to 'testcase-comparison-2605'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Alright.  Fixed in the most straightforward way, I think.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix set owner to @exarkun|
|-|-|

Even though this fixes another test, I think this change should have some new unit tests associated with it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25212]) test the comparison stuff directly

refs [#2605](https://github.com/twisted/twisted/issues/2605)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Alright

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_9"></a>|@radix set owner to @exarkun|
|-|-|

Ok, looks good. +1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [25227]) Merge testcase-comparison-2605

Author: exarkun
Reviewer: radix
Fixes: [#2605](https://github.com/twisted/twisted/issues/2605)

Explicitly define equal and not-equal comparison for `TestCase` so that instances
only compare equal to themselves.  This is the Python 2.3 - 2.5 behavior but not
the Python 2.6 behavior.  It is required in order for a particular test method to
be runnable more than once in a single suite.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
